modules/pubCommonIdSystem.js

```diff
@@ -46,52 +140,78 @@ export const pubCommonIdSubmodule = {
    * @function
    * @param {SubmoduleConfig} [config]
    * @returns {IdResponse}
    */
-  getId: function ({params: {create = true, pixelUrl} = {}} = {}) {
+  getId: function (config = {}) {
+    const {params: {create = true, pixelUrl} = {}} = config;
+    let newId;
     try {
       if (typeof window[PUB_COMMON_ID] === 'object') {
         // If the page includes its own pubcid module, then save a copy of id.
-        return {id: window[PUB_COMMON_ID].getId()};
+        newId = window[PUB_COMMON_ID].getId();
       }
     } catch (e) {
     }
 
-    const newId = (create && utils.hasDeviceAccess()) ? utils.generateUUID() : undefined;
+    if (!newId) newId = (create && utils.hasDeviceAccess()) ? utils.generateUUID() : undefined;
     return {
       id: newId,
-      callback: this.makeCallback(pixelUrl, newId)
+      callback: getIdCallback(newId, this.makeCallback(pixelUrl, newId), config)
     }
   },
   /**
-   * performs action to extend an id
+   * performs action to extend an id.  There are generally two ways to extend the expiration time
+   * of stored id: using pixelUrl or return the id and let main user id module write it again with
+   * the new expiration time.
+   *
+   * PixelUrl, if defined, should point back to a first party domain endpoint.  On the server
+   * side, there is either a plugin, or customized logic to read and write back the pubcid cookie.
+   * The extendId function itself should return only the callback, and not the id itself to avoid
+   * having the script-side overwriting server-side.  This applies to both pubcid and sharedid.
+   *
+   * On the other hand, if there is no pixelUrl, then the extendId should return storedId so that
+   * its expiration time is updated.  Sharedid, however, will have to be updated by this submodule
+   * separately.
+   *
    * @function
    * @param {SubmoduleParams} [config]
    * @param {Object} storedId existing id
    * @returns {IdResponse|undefined}
    */
-  extendId: function({params: {extend = false, pixelUrl} = {}} = {}, storedId) {
-    try {
-      if (typeof window[PUB_COMMON_ID] === 'object') {
-        // If the page includes its onw pubcid module, then there is nothing to do.
-        return;
-      }
-    } catch (e) {
-    }
+  extendId: function(config = {}, storedId) {
+    const {params: {extend = false, pixelUrl} = {}} = config;
 
     if (extend) {
-      // When extending, only one of response fields is needed
-      const callback = this.makeCallback(pixelUrl, storedId);
-      return callback ? {callback: callback} : {id: storedId};
+      try {
+        if (typeof window[PUB_COMMON_ID] === 'object') {
+          // If the page includes its own pubcid module, then there is nothing to do
+          // except to update sharedid
+          storeData(config, readData(config));
+          return;
+        }
+      } catch (e) {
+      }
+
+      if (pixelUrl) {
+        const callback = this.makeCallback(pixelUrl, storedId);
```

Comment:
Should this be similar to callback in getId()? The getId() has callback: getIdCallback(newId, this.makeCallback(pixelUrl, newId), config) where the shared id (
_sharedid) cookie is stored if shared id is not opted out.
Based on the shared id opt out, should the cookie(_sharedid) be reset/removed in extendId() function?

Reply:
You have good point.  What it means is that shareid.org should still be called in extendId(), and if it returns an opt-out, then erase the stored value.

Reply:
@SKOCHERI After checking the flow a bit, extendId is going to be called at every page load, so it seems a bit heavy handed. There is an actually a superior mechanism with refreshInSeconds that can be used to check the validity of the stored sharedid.  That would drive getid() periodically, then we can modify the existing callback to delete the cookie when opted out.

Reply:
Yeah. Basically the cookie needs to be cleared on shared id opt out. refreshInSeconds would also work.

Reply:
Thanks.  Good catch.